lib/plugins/aws/package/compile/functions/index.js

```diff
@@ -730,4 +754,39 @@ class AwsCompileFunctions {
   }
 }
 
+function addFileContentsToHashes(filePath, hashes) {
+  return new BbPromise((cb, reject) => {
+    const readStream = fs.createReadStream(filePath);
+    readStream
+      .on('data', chunk => {
+        hashes.forEach(hash => {
+          hash.write(chunk);
+        });
+      })
+      .on('close', () => {
+        cb();
+      })
+      .on('error', error => {
+        reject(new Error(`Could not add file content to hash: ${error}`));
+      });
+  });
+}
+
+function extractLayerConfigurationsFromFunction(functionProperties, cfTemplate) {
+  const layerConfigurations = [];
+  if (!functionProperties.Layers) return layerConfigurations;
+  functionProperties.Layers.forEach(potentialLocalLayerObject => {
+    if (potentialLocalLayerObject.Ref) {
+      const configuration = _.cloneDeep(cfTemplate.Resources[potentialLocalLayerObject.Ref]);
```

Comment:
Main reason was I assumed this line https://github.com/pwithams/serverless/blob/c8be293c97dd8550f2a5d02ee8d61e6032da3335/lib/plugins/aws/package/compile/functions/index.js#L513 would delete part of the already created cloudformation template if the reference was returned.

Also, in general it seemed like `cfTemplate` is such an important object that if part of it is being returned then it should be a copy so that anything using the return value doesn't accidentally alter it.

However, if the delete doesn't actually affect anything (I think removing the clone only fails one of my tests?) then it can be taken out, especially if lodash usage is being minimized.

Reply:
> `cfTemplate` is such an important object that if part of it is being returned then it should be a copy so that anything using the return value doesn't accidentally alter it.

In fact, if we're taking it as reference, then indeed we should never attempt to modify it. However implying a practice that whenever we take it, we copy it (just in case), could be noisy and difficult to follow, it's definitely not natural in JS.
It's more natural to purely stay away from modifying such reference objects.

So I'd say, let's not clone and let's ensure we do not modify it in further processing

Reply:
That makes sense. I think it still has to be modified later in some way as `S3Key` can't be part of the hash otherwise it will change every time. I've now moved the clone out of the function and applied it to the returned result instead - this solves the `configuration._serverlessLayerName` naming issue below too.

Let me know if you have any other ideas how to handle keeping `S3Key` out of the hash without a clone.